source-mongodb/pull.go

```diff
@@ -183,83 +148,118 @@ type resourceState struct {
 }
 
 type docKey struct {
 	Id interface{} `bson:"_id"`
 }
+type namespace struct {
+	Database string `bson:"db"`
+	Collection string `bson:"coll"`
+}
 type changeEvent struct {
-	DocumentKey   docKey `bson:"documentKey"`
-	OperationType string `bson:"operationType"`
-	FullDocument  bson.M `bson:"fullDocument"`
+	DocumentKey   docKey    `bson:"documentKey"`
+	OperationType string    `bson:"operationType"`
+	FullDocument  bson.M    `bson:"fullDocument"`
+	Ns            namespace `bson:"ns"`
 }
 
 const resumePointGoneErrorMessage = "the resume point may no longer be in the oplog"
 const resumeTokenNotFoundErrorMessage = "the resume token was not found"
 
-func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, binding int, res resource, state *resourceState) error {
-	var db = client.Database(res.Database)
+func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, resources []resource, state captureState) error {
+	if len(resources) < 1 {
+		return nil
+	}
 
-	var collection = db.Collection(res.Collection)
+	// If we have a stream resume token, we use that to continue our change stream
+	// otherwise, we look at BackfillStartedAt of all resources and use the
+	// minimum value to start our change stream from
+	var streamStartAt time.Time
+	if len(state.StreamResumeToken) < 1 {
+		for _, res := range resources {
+			var resState = state.Resources[resourceId(res)]
+
+			if !resState.BackfillStartedAt.IsZero() {
+				if streamStartAt.IsZero() {
+					streamStartAt = resState.BackfillStartedAt
+				} else if resState.BackfillStartedAt.Before(streamStartAt) {
+					streamStartAt = resState.BackfillStartedAt
+				}
+			}
+		}
+	}
+
+	var collectionFilters = make([]bson.D, len(resources))
+	var collectionBindingIndex = make(map[string]int)
+	for idx, res := range resources {
+		collectionBindingIndex[resourceId(res)] = idx
+
+		collectionFilters[idx] = bson.D{
+			{"ns.db", res.Database},
+			{"ns.coll", res.Collection},
+		}
+	}
 
 	log.WithFields(log.Fields{
-		"database":           res.Database,
-		"collection":         res.Collection,
-		"state":              state,
-	}).Info("listening on changes on collection")
-	var eventFilter = bson.D{{"$match", bson.D{{"$or",
-	bson.A{
-		bson.D{{"operationType", "delete"}},
-		bson.D{{"operationType", "insert"}},
-		bson.D{{"operationType", "update"}},
-		bson.D{{"operationType", "replace"}},
-	}}}}}
+		"state":             state,
+		"collectionFilters": collectionFilters,
+		"streamResumeToken": state.StreamResumeToken,
+		"streamStartAt":     streamStartAt,
+	}).Info("listening on changes")
+
+	var eventFilter = bson.D{{"$match", bson.D{{
+		"$and", bson.A{
+			bson.D{{"$or",
+				bson.A{
+					bson.D{{"operationType", "delete"}},
+					bson.D{{"operationType", "insert"}},
+					bson.D{{"operationType", "update"}},
+					bson.D{{"operationType", "replace"}},
+				},
+			}},
+		  bson.D{{"$or", collectionFilters}},
```

Comment:
It would be really nice to get _all_ events from all collections of all types. That way we have the most up to date resume cursor possible - I'm thinking about if there was a source where we were capturing only a single collection of many, and that collection had a low rate of changes compared to others and the resume token we got from that single collection ends up getting pushed out of the oplog. This would then require filtering out oplog events we don't care about in the connector. I believe SQL captures do something like this. Would that be practical here?

Reply:
I see your point, I personally wante to avoid touching data that we didn't need to touch, hence the filtering, but I can see that it has implications for our connector being able to work in certain scenarios, that should work 👍🏽